Re-add: Use content type to set extension for downloaded files

SergioEstevao opened the pull request:
Re-add [fix to use content type to set file extension for audio files](https://github.com/Automattic/pocket-casts-ios/pull/1577)

I needed to re-add the columns on episode tables for this fix to work correctly.

I also updated the code use an existing MetadataTask class to update the contentType instead of doing it on the end of the upload.

In order to make sure this change didn't affect performance in queries because of Episode table changes I run some performance measurement using the Filters tab and measuring the method: `episodeCount(forFilter: EpisodeFilter, episodeUuidToAdd: String?)`

Here are the results I got [before and after these changes](https://docs.google.com/spreadsheets/d/1n16l_q1j-DKrk_kBxJIrbJFUYX1YRHv3vhHq8lV60kU/edit?usp=sharing) on a iPhone Xs

| Before | After |
| - | - |
|  0.3064338418  |   0.3009749147 | 

## To test

## Test if the fix worked

- Start the app on the simulator
- Search for the following podcast: `ondacer` on the discovery section
- Subscribe to the podcast
- Open the podcast and download one of the episodes
- Tap play
- The episode should play correctly
- Now play an episode without downloading it first and see if it works correctly
- Smoke test other podcasts, to see if this change doesn't brake anything on that front

## Test DB migrations because of fields `contentType` being added again

- Install a 7.62 build then migrate to this version 7.66
- Install a 7.61 build then migrate to this version 7.66
- Install a 7.64 build then migrate to this version 7.66

## Test the performance of the app because of database changes

- Setup the app to use a large database by cherry picking the last 4 commits in this branch: `hang/copy_database`
- Use the large user database with 500mb
- Use a real device
- Start the app
- Go to filters view
- Scroll around

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

SergioEstevao committed:
Use contentType to set extension for downloaded files

Some podcasts use file URLs with an extension that does not match with the contentType of the files.
This changes saves the content file provided by the header response from servers and use that to use as an extension

# Conflicts:
#	Modules/DataModel/Sources/PocketCastsDataModel/Private/Managers/Util/DatabaseHelper.swift

SergioEstevao committed:
Add back content type column on episodes.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-ios/milestone/47">7.64</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

leandroalonso reviewed (changes requested):
@SergioEstevao I'm not sure we should re-add columns to the `SJEpisode` table.

If we want to follow this path, I think we need to answer if the query time average changes for big databases by adding those columns. We should probably do some sort of average time or something like that.

Alternatively, we could store this somewhere.

bjtitus commented:
I think this is a separate item from this PR but should we switch from `FileTypeUtil` to [`UTType(mimeType:...)`](https://developer.apple.com/documentation/uniformtypeidentifiers/uttype) for parsing the `Content-Type` header and [`UTType.preferredFilenameExtension`](https://developer.apple.com/documentation/uniformtypeidentifiers/uttype) to fetch the extension at some point?

SergioEstevao committed:
Merge branch 'trunk' into re-add/use-content-type-to-set-extension-for-downloaded-files

SergioEstevao committed:
Simplify code to update content type to use existing MetadataTask

SergioEstevao committed:
Move measure time to utils class

SergioEstevao committed:
Measure time of query for counts

SergioEstevao committed:
Move back the drop column to normal step

If we don't have it here the DataManager can for certain version run the cleanUp method after and drop the column after we just created it.

SergioEstevao committed:
Remove empty space

SergioEstevao commented on Modules/DataModel/Sources/PocketCastsDataModel/Private/Managers/Util/DatabaseHelper.swift:
```diff
@@ -717,11 +717,29 @@ class DatabaseHelper {
         if schemaVersion < 48 {
             schemaVersion = 48
+            do {
+                try db.executeUpdate("ALTER TABLE SJEpisode DROP COLUMN contentType", values: nil)
```
I'm not sure if we should put this back here but my rational is:
- Some users maybe be on versions where where the  clean up process is not run, so they will have the new columns, so we need to drop it and then create it again on version 50.
- The reason for the above is that cleanUp code is run after this update code.

leandroalonso replied:
I think what we might be able to do is to just remove the `DROP TABLE` from everywhere.

Then on `50` you can do:

```
try? db.executeUpdate("ALTER TABLE SJEpisode ADD COLUMN contentType TEXT;", values: nil)
try? db.executeUpdate("ALTER TABLE SJUserEpisode ADD COLUMN contentType TEXT;", values: nil)
```

If the column exists, this will fail but we're using `try?`. If not, it will be created. This way we avoid the unnecessary work of removing and re-adding for those users.

I have to test this though.

SergioEstevao replied:
@leandroalonso good suggestion, just changed the code to use try?.

SergioEstevao commented on Modules/DataModel/Sources/PocketCastsDataModel/Public/DataManager.swift:
```diff
@@ -849,7 +843,12 @@ public class DataManager {
 
     public func episodeCount(forFilter: EpisodeFilter, episodeUuidToAdd: String?) -> Int {
-        filterManager.episodeCount(forFilter: forFilter, episodeUuidToAdd: episodeUuidToAdd, dbQueue: dbQueue)
+        var count = 0
```
This should be reverted after the tests.

SergioEstevao committed:
Change update to fail silently if column already exists

SergioEstevao committed:
Update changelog entries

leandroalonso reviewed (approved):
Tested on the following scenarios:

✅ Version without the column. Column is created and the downloaded file plays correctly;
✅ Version with the column. Column was still there and the downloaded file plays correctly
✅ Pretty old version without the column. Column is created and the downloaded file plays correctly;

SergioEstevao committed:
Remove measuring of time.
